pkg/plugin/plugin.go

```diff
@@ -296,10 +391,79 @@ func (g *GlooMatchedRouteTable) matchRoutes(logCtx *logrus.Entry, rollout *v1alp
 					CanaryOrPreviewDestination: canary,
 				},
 			}
-			logCtx.Debugf("adding destination %+v", dest)
 			g.HttpRoutes = append(g.HttpRoutes, dest)
 		}
 	} // end range httpRoutes
 
 	return nil
 }
+
+func buildGlooHTTPRoute(name string, matcher *solov2.HTTPRequestMatcher, canaryDestination *SimpleDestinationReference) *GlooMatchedHttpRoutes {
+	return &GlooMatchedHttpRoutes{
+		HttpRoute: &networkv2.HTTPRoute{
+			Name: name,
+			Matchers: []*solov2.HTTPRequestMatcher{
+				matcher,
+			},
+			ActionType: &networkv2.HTTPRoute_ForwardTo{
+				ForwardTo: &networkv2.ForwardToAction{
+					Destinations: []*solov2.DestinationReference{
+						{
+							RefKind: &solov2.DestinationReference_Ref{
+								Ref: &solov2.ObjectReference{
+									Name:      canaryDestination.Reference.Name,
+									Namespace: canaryDestination.Reference.Namespace,
+								},
+							},
+							Port: &solov2.PortSelector{
+								Specifier: &solov2.PortSelector_Number{
+									Number: canaryDestination.Port.Number,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func buildGlooMatches(headerRouting *v1alpha1.SetHeaderRoute) *solov2.HTTPRequestMatcher {
+	matcher := &solov2.HTTPRequestMatcher{
+		Name:    headerRouting.Name + "-matcher",
+		Headers: []*solov2.HeaderMatcher{},
+	}
+
+	for _, m := range headerRouting.Match {
+		var isRegex bool
+		var matchValue string
+		if m.HeaderValue.Exact != "" {
+			matchValue = m.HeaderValue.Exact
+		} else if m.HeaderValue.Regex != "" {
+			matchValue = m.HeaderValue.Regex
+			isRegex = true
+		} else if m.HeaderValue.Prefix != "" {
+			// error, do nothing?
+			// try to convert to a regex for the user?
+			panic("TODO: impl some handling of prefix")
+		}
+		headerMatcher := &solov2.HeaderMatcher{
+			Name:  m.HeaderName,
+			Value: matchValue,
+			Regex: isRegex,
+		}
```

Comment:
What happens if no `Exact`, `Regex` or `Prefix` is specified?

Reply:
golang zero value things. non-regex matching on "" as the value. 🤔

Reply:
ok, real answer. Argo rollouts itself errors with this message https://github.com/argoproj/argo-rollouts/blob/1264f8d9b31c53177b666a9145bdb02476f271ad/pkg/apis/rollouts/validation/validation.go#L44

and our plugin doesn't seem to get invoked at all.